src/kaleidoscope/hid.cpp

```diff
@@ -51,13 +51,86 @@
 namespace kaleidoscope {
 namespace hid {
 
+// Anonymous namespace for internal variables and helper functions
+namespace {
+
+// The functions in this namespace are primarily to solve the problem of rollover from a
+// key with a modifier flag (e.g. `LSHIFT(Key_T)`) to one without (e.g. `Key_H`), which
+// used to result in the mod flag being applied to keys other than the one with the
+// flag. By using `mod_flags_allowed`, we can prohibit any modifier flags that aren't in
+// the key most recently pressed, fixing the rollover problem, and getting the intended
+// `The` instead of `THe`. This was particularly noticeable to people who were using a
+// separate layer for `shift`, and a OneShot layer key to activate it, causing what looked
+// like a OneShot bug.
+
+// A bitfield of the current modifier flags from keys processed so far in the current
+// keyswitch scan:
+static byte current_mod_flags{0};
+
+// The current set of modifier flags that will be allowed to be added to the upcoming
+// keyboard HID report. This should be set by the latest key pressed:
+static byte mod_flags_allowed{0};
+
+// The keycode of the last key to toggle on. This get
+static byte toggled_on_keycode{0};
+
+// This returns true if the key is a keyboard key, and its keycode is a modifier.
+bool isPureModifier(Key mappedKey) {
+  // If it's not a keyboard key, return false
+  if (mappedKey.flags & (SYNTHETIC | RESERVED))
+    return false;
+  // An efficient way to check if the keycode is for a modifier. Key_LeftControl has the
+  // modifier keycode with the lowest value, and the others are sequential, and we use an
+  // unsigned integer so after subtraction, the comparison is simple.
+  byte offset_keycode = mappedKey.keyCode - Key_LeftControl.keyCode;
```

Comment:
This idiom is different than the code we use elsewhere in Kaleidoscope. If it actually generates tighter assembly code, that's great. 

Either way, it should get factored out into a method named something like "isModifier()"

This is also a case where I'd prefer "uint8_t" to "byte" - With a typename that explicitly declares that it's an unsigned integer, it's obvious that we won't overflow. "byte" doesn't make that as clear.

Reply:
I've used `byte` so much now that it's very clear to _me_ that it means `unsigned char`, but I have no objection to changing it to `uint8_t` if you prefer that.

Reply:
When I wrote this, I was remembering coming up with this trick in a context where I wanted to check if a keycode was a modifier, and if so, use the modifier-byte index number. For that, it was more efficient to do the subtraction first, then use the resulting number to do a bit shift, but if all we need to do is the test for a modifier keycode, the resulting code is the same size, so there's no reason to do so here. I'll update the PR to use the same idiom as in KeyboardioHID.